From 9e555d835928ed9c36ce030fe5b49f8fbf224018 Mon Sep 17 00:00:00 2001
From: Charles Bousseau <cbousseau@anaconda.com>
Date: Sun, 20 Jul 2025 14:03:26 -0400
Subject: [PATCH] metal gpu selection

In macOS, in order for the system to provide a default Metal device object, you must link to the Core Graphics framework.
You usually need to do this explicitly if you're writing apps that don't use graphics by default, such as command line tools.
https://developer.apple.com/documentation/metal/1433401-mtlcreatesystemdefaultdevice?language=objc
Systems with Apple silicon only have one GPU, which removes the need to choose a GPU.
https://developer.apple.com/documentation/metal/mtldevice/1433409-lowpower#discussion

I did try linking to CoreGraphics, but MTLCreateSystemDefaultDevice was still returning nil.

Updated for b8994. File rename from ggml/src/ggml-metal/ggml-metal.m to ggml-metal-device.m happened in b6653.
---
 ggml/src/ggml-metal/ggml-metal-device.m | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/ggml/src/ggml-metal/ggml-metal-device.m b/ggml/src/ggml-metal/ggml-metal-device.m
index d1bce8b..046441b 100644
--- a/ggml/src/ggml-metal/ggml-metal-device.m
+++ b/ggml/src/ggml-metal/ggml-metal-device.m
@@ -1116,6 +1116,25 @@ ggml_metal_device_t ggml_metal_device_init(int device, int n_devices) {
     @autoreleasepool {
         if (dev->mtl_device == nil) {
             dev->mtl_device = MTLCreateSystemDefaultDevice();
+            if (dev->mtl_device == nil) {
+              /*
+                In macOS, in order for the system to provide a default Metal device object, you must link to the Core Graphics framework.
+                You usually need to do this explicitly if you're writing apps that don't use graphics by default, such as command line tools.
+                > https://developer.apple.com/documentation/metal/1433401-mtlcreatesystemdefaultdevice?language=objc
+                Systems with Apple silicon only have one GPU, which removes the need to choose a GPU.
+                > https://developer.apple.com/documentation/metal/mtldevice/1433409-lowpower#discussion
+               */
+                NSArray * devices = MTLCopyAllDevices();
+                for (id<MTLDevice> dev_tmp in devices) {
+                    if (dev_tmp != nil) {
+                        if (dev->mtl_device == nil) {
+                            dev->mtl_device = dev_tmp;
+                        } else {
+                            [dev_tmp release];
+                        }
+                    }
+                }
+            }
 
             if (dev->mtl_device) {
                 dev->mtl_queue = [dev->mtl_device newCommandQueue];
